Fix for seekTo feature

a1shadows opened the pull request:
# <Feature Title>

Fixes #631 

PR Branch
**_#ADD LINK TO THE PR BRANCH_**

## Motivation and Context
If there are partitions for which there is no valid offset corresponding to the timestamp to seek to, the offsetForTimestamp method still includes it in the Map returned, but with a NULL value. We need to skip all partitions that have offset as NULL when subscribing to avoid an error when we try to seek to the offset at a later step.

## Checklist:

* [ ] Unit tests added

* [ ] Integration tests added

* [ ] Test names are meaningful

* [ ] Feature manually tested

* [ ] Branch build passed

* [ ] No 'package.*' in the imports

* [ ] Relevant Wiki page updated with clear instruction for the end user
  * [ ] Not applicable. This was only a refactor change, no functional or behaviour changes were introduced

* [ ] Http test added to `http-testing` module(if applicable) ?
  * [ ] Not applicable. The changes did not affect HTTP automation flow

* [ ] Kafka test added to `kafka-testing` module(if applicable) ?
  * [ ] Not applicable. The changes did not affect Kafka automation flow

a1shadows commented:
@authorjapps  There was a bug in the seekTo feature . In situations where there are multiple partitions on the topic and some of the partitions do not have a valid offset corresponding the timestamp to seek to, the seek operation will give an NPE.

Can you review the fix?

authorjapps commented:
> @authorjapps There was a bug in the seekTo feature . In situations where there are multiple partitions on the topic and some of the partitions do not have a valid offset corresponding the timestamp to seek to, the seek operation will give an NPE.
> 
> Can you review the fix?

Got you @a1shadows . That scenario causing NPE.
Is there anyway to reproduce this in a single node Kafka in our docker setup?

e.g.
by creating a topic with two or three partitions, then sending records to 1 and 3 based on their key, then not sending any record to 2.
Then try to consume with seekTimestamp, it should fail with NPE

Then with the fix, it should pass as expected(without NPE)

a1shadows commented:
>  e.g.
> by creating a topic with two or three partitions, then sending records to 1 and 3 based on their key, then not sending any record to 2.
> Then try to consume with seekTimestamp, it should fail with NPE
> 
> Then with the fix, it should pass as expected(without NPE)

@authorjapps  I'll look into this

a1shadows commented:
@authorjapps  I've added a multi partition topic test to the CI. For this, I added an init container to the docker compose file, since by default the cp-kafka docker image does not accept num.partitions as an env variable for auto created topics. I ran it locally and it ran fine. However, for the CI the docker compose files are picked from the master branch so can't get a proper test on here

Do you have anything in mind for this? Can we run the CI using the docker compose files on the branch we are trying to merge?

a1shadows commented on .github/workflows/main.yml:
```diff
@@ -18,9 +18,7 @@ jobs:
-        run: wget https://raw.githubusercontent.com/authorjapps/zerocode-docker-factory/master/compose/kafka-schema-registry.yml
       - name: Running Kafka
-        run: docker-compose -f kafka-schema-registry.yml up -d && sleep 10
+        run: docker-compose -f docker/compose/kafka-schema-registry.yml up -d && sleep 10
```
@authorjapps  hey these are the changes to run the docker compose file present in the repo. What are your thoughts on these changes?

authorjapps replied:
> @authorjapps hey these are the changes to run the docker compose file present in the repo. What are your thoughts on these changes?

@a1shadows , Looks good to me. Did you check whether CI picked this command?

a1shadows replied:
@authorjapps  Yes, as per the CI logs, can see this ran successfully:
```
Run docker-compose -f docker/compose/kafka-schema-registry.yml up -d && sleep 10
```

can see the init container also started in the logs:
```
Creating compose_init-kafka-container_1 ... done
```

a1shadows commented on docker/compose/kafka-schema-registry.yml:
```diff
@@ -21,6 +21,22 @@ services:
       KAFKA_INTER_BROKER_LISTENER_NAME: PLAINTEXT
       KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
 
+  init-kafka-container:
```
@authorjapps  have added an init container that will create a multi partition topic to run the test against.

Ideally we should change the `num.partitions` setting in the kafka broker, but can't find a way to do that in the cp-kafka image we have used. Let me know your thoughts

authorjapps replied:
> @authorjapps have added an init container that will create a multi partition topic to run the test against.
> 
> Ideally we should change the `num.partitions` setting in the kafka broker, but can't find a way to do that in the cp-kafka image we have used. Let me know your thoughts


ok. Let's figure out this in the coming days.

Also,
if you have tested locally in your Kafka cluster that everything works as expected, then we should be fine for now. 
I will raise a TECH-DEBT to get a Integration test added to our CI suite to cover this `PARTITION->OFFSET NPE` scenario.

Let's get this merged anyways.

authorjapps commented on .github/workflows/main.yml:
```diff
@@ -18,9 +18,7 @@ jobs:
-        run: wget https://raw.githubusercontent.com/authorjapps/zerocode-docker-factory/master/compose/kafka-schema-registry.yml
       - name: Running Kafka
-        run: docker-compose -f kafka-schema-registry.yml up -d && sleep 10
+        run: docker-compose -f docker/compose/kafka-schema-registry.yml up -d && sleep 10
```
👍
